api/services/control/control.proto

```diff
@@ -163,3 +167,56 @@ message InfoRequest {}
 message InfoResponse {
 	moby.buildkit.v1.types.BuildkitVersion buildkitVersion = 1;
 }
+
+message BuildHistoryRequest {
+	bool ActiveOnly = 1;
+	string Ref = 2;
+}
+
+enum BuildHistoryEventType {
+	STARTED = 0;
+	COMPLETE = 1;
+	DELETED = 2;
+}
+
+
+message BuildHistoryEvent {
+	BuildHistoryEventType type = 1;
+	BuildHistoryRecord record = 2;
+}
+
+message BuildHistoryRecord {
+	string Ref = 1;
+	string Frontend = 2;
+	map<string, string> FrontendAttrs = 3;
+	repeated Exporter Exporters = 4;
+	google.rpc.Status error = 5;
+	google.protobuf.Timestamp CreatedAt = 6 [(gogoproto.stdtime) = true];
+	google.protobuf.Timestamp CompletedAt = 7 [(gogoproto.stdtime) = true];
+	Descriptor logs = 8;
+	map<string, string> ExporterResponse = 9;
+	BuildResultInfo Result = 10;
+	map<string, BuildResultInfo> Results = 11;
+	int32 Generation = 12;
+	// TODO: tags
+	// TODO: steps/cache summary
+	// TODO: unclipped logs
+	// TODO: pinning
+}
+
+message Descriptor {
```

Comment:
@AkihiroSuda I couldn't figure out how to import this type from containerd's proto definitions and not break compilation. Maybe you have better ideas on how to make it work so this duplication is not needed. I think the issue is that containerd uses `Protobuild` with some override mapping for import rules?

Reply:
I think duplicating the `Descriptor` definition is fine.
The duplicated amount is small, and the definition is unlikely going to be changed in a decade.

Also, I guess we can dedupe this later without breaking the compatibility?

Reply:
This also means that we would need to convert between different variants of descriptor on create and parse. To actually download the blob I'm planning to just enable the contentstore methods in control api. This means that either I will run into the same issue or some of the API endpoints will use one descriptor type and some the other.